core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/hashJoin/HashJoinOpExecConfig.scala

```diff
@@ -0,0 +1,103 @@
+package edu.uci.ics.texera.workflow.operators.hashJoin
+
+import akka.actor.ActorRef
+import akka.util.Timeout
+import edu.uci.ics.amber.engine.architecture.breakpoint.globalbreakpoint.GlobalBreakpoint
+import edu.uci.ics.amber.engine.architecture.controller.Workflow
+import edu.uci.ics.amber.engine.architecture.deploysemantics.deploymentfilter.UseAll
+import edu.uci.ics.amber.engine.architecture.deploysemantics.deploystrategy.RoundRobinDeployment
+import edu.uci.ics.amber.engine.architecture.deploysemantics.layer.WorkerLayer
+import edu.uci.ics.amber.engine.architecture.worker.WorkerState
+import edu.uci.ics.amber.engine.common.Constants
+import edu.uci.ics.amber.engine.common.amberexception.WorkflowRuntimeException
+import edu.uci.ics.amber.engine.common.ambertag.{AmberTag, LayerTag, OperatorIdentifier}
+import edu.uci.ics.amber.engine.common.tuple.ITuple
+import edu.uci.ics.amber.engine.operators.OpExecConfig
+import edu.uci.ics.amber.error.WorkflowRuntimeError
+import edu.uci.ics.texera.workflow.common.operators.OperatorExecutor
+import edu.uci.ics.texera.workflow.common.operators.source.SourceOpExecConfig
+import edu.uci.ics.texera.workflow.common.tuple.Tuple
+
+import scala.collection.mutable
+import scala.concurrent.ExecutionContext
+
+class HashJoinOpExecConfig(
+    override val tag: OperatorIdentifier,
+    val opExec: Int => OperatorExecutor,
+    val probeAttribute: String,
+    val buildAttribute: String
+) extends OpExecConfig(tag) {
+
+  var buildTableTag: LayerTag = _
+
+  override lazy val topology: Topology = {
+    new Topology(
+      Array(
+        new WorkerLayer(
+          LayerTag(tag, "main"),
+          opExec,
+          Constants.defaultNumWorkers,
+          UseAll(),
+          RoundRobinDeployment()
+        )
+      ),
+      Array(),
+      Map()
+    )
+  }
+
+  private def getBuildTableOpIdentifier(): OperatorIdentifier = {
+    var buildOpId: Option[OperatorIdentifier] =
+      inputToOrdinalMapping.keys.find(opId => (inputToOrdinalMapping(opId) == 0))
+    buildOpId match {
+      case Some(opId) => return opId
+      case None =>
+        val error = WorkflowRuntimeError(
+          "No operator identifier has input num 0",
+          "HashJoinOpExecConfig",
+          Map()
+        )
+        opExecConfigLogger.logError(error)
+        throw new WorkflowRuntimeException(error)
+    }
+  }
+
+  override def runtimeCheck(
+      workflow: Workflow
+  ): Option[mutable.HashMap[AmberTag, mutable.HashMap[AmberTag, mutable.HashSet[LayerTag]]]] = {
+    assert(workflow.inLinks(tag).nonEmpty)
+    buildTableTag = workflow.operators(getBuildTableOpIdentifier()).topology.layers.last.tag
```

Comment:
is runtimeCheck still needed?

Reply:
Yes. It decides which operator to start first and then controller sends start to it. After that controller sends start to others scans and there is some dependency graph that is maintained in controller. I suggest we keep it till we refactor controller side.